### PR TITLE
Engine abstraction remove legacy writer

### PR DESCRIPTION
**Pull request by pihme**
## Description

- Introduces a new task that can be scheduled by the engine and that can return records to be written
- Replaces all references to `LegacyTypedStreamWriter`outside of stream processor with scheduling of the new tasks
- The overall result is that the engine no longer depends on any writer that writes directly to the stream. Stream processor is now in full control of when and how records shall be written (Heureka!)

## Related issues

related to #9724, #9730, #9725

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [X] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

File: engine/src/main/java/io/camunda/zeebe/engine/api/TaskResult.java
Comment (Zelldon) [inline engine/src/main/java/io/camunda/zeebe/engine/api/TaskResult.java:18]: Yes I think this wouldn't be the end solution, since I would like to hide the writer completely but it is a great and pragmatic step forward :) :+1:

File: engine/src/main/java/io/camunda/zeebe/engine/api/TaskResultBuilder.java
Comment (Zelldon) [inline engine/src/main/java/io/camunda/zeebe/engine/api/TaskResultBuilder.java:28]: :wrench: I feel that we don't need them here. I think they will always set to default values, since we will not write rejections here.

File: engine/src/main/java/io/camunda/zeebe/engine/api/ProcessingScheduleService.java
Comment (Zelldon) [inline engine/src/main/java/io/camunda/zeebe/engine/api/ProcessingScheduleService.java:34]: :thinking: Didn't we removed that on purpose why we have that again?

File: engine/src/main/java/io/camunda/zeebe/engine/api/ProcessingScheduleService.java
Comment (Zelldon) [inline engine/src/main/java/io/camunda/zeebe/engine/api/ProcessingScheduleService.java:16]: :question: Can we remove this now?

File: engine/src/main/java/io/camunda/zeebe/engine/api/Task.java
Comment (Zelldon) [inline engine/src/main/java/io/camunda/zeebe/engine/api/Task.java:12]: :wrench: Maybe we add here also an comment that this is the intermediate solution and ideally the builder is in the end created in the ScheduleService consumer classes

File: engine/src/main/java/io/camunda/zeebe/streamprocessor/DirectTaskResultBuilder.java
Comment (Zelldon) [inline engine/src/main/java/io/camunda/zeebe/streamprocessor/DirectTaskResultBuilder.java:24]: :thought_balloon: First thought we could also reuse the other build if it just implements this interface as well, but I guess this implementation is also simple enough :+1:

File: engine/src/main/java/io/camunda/zeebe/streamprocessor/ProcessingScheduleServiceImpl.java
Comment (Zelldon) [inline engine/src/main/java/io/camunda/zeebe/streamprocessor/ProcessingScheduleServiceImpl.java:49]: :question: I'm a bit confused, which method is called here. Is this a loop?

Comment (Zelldon) [reply engine/src/main/java/io/camunda/zeebe/streamprocessor/ProcessingScheduleServiceImpl.java:49]: Ah I see there is a default implementation in the interface :sweat_smile:

**Review by ChrisKujawa — APPROVED** [review]
@pihme didn't expected a PR regarding this :grin: but this a significant step forward. Awesome thank you :trophy: :hugs: 

I had some comments, please see below before merging :)

File: engine/src/main/java/io/camunda/zeebe/streamprocessor/DirectTaskResultBuilder.java
Comment (pihme) [reply engine/src/main/java/io/camunda/zeebe/streamprocessor/DirectTaskResultBuilder.java:24]: I think you will want to overhaul this anyway later. So I thought a bit of code duplication doesn't hurt, to better see what is used where.

File: engine/src/main/java/io/camunda/zeebe/engine/api/ProcessingScheduleService.java
Comment (pihme) [reply engine/src/main/java/io/camunda/zeebe/engine/api/ProcessingScheduleService.java:16]: I think the whole interface needs an overhaul. I pretty much added methods so I could migrate existing code. And then I was hoping you would look at this mess and clean up the interface at some point.

File: engine/src/main/java/io/camunda/zeebe/streamprocessor/ProcessingScheduleServiceImpl.java
Comment (pihme) [reply engine/src/main/java/io/camunda/zeebe/streamprocessor/ProcessingScheduleServiceImpl.java:53]: see comment above on quality level of this class

File: engine/src/main/java/io/camunda/zeebe/engine/api/ProcessingScheduleService.java
Comment (Zelldon) [reply engine/src/main/java/io/camunda/zeebe/engine/api/ProcessingScheduleService.java:16]: Alright I will do this then :) :+1:

File: engine/src/main/java/io/camunda/zeebe/engine/api/ProcessingScheduleService.java
Comment (pihme) [reply engine/src/main/java/io/camunda/zeebe/engine/api/ProcessingScheduleService.java:34]: see other comment on quality level of this class

**Comment by pihme** [comment]
bors merge

**Comment by pihme** [comment]
bors retry

**Comment by Zelldon** [comment]
bors r- 

@pihme I need to take a look at the scheduling first I think there might be some issue with the processing and scheduling work in between

**Comment by Zelldon** [comment]
I will merge it after I have checked it :)

**Review by ChrisKujawa — COMMENTED** [review]
Hey @pihme 

I added some more commits to your PR I hope that was OK. This allows us to make sure that the tasks are only scheduled between processing, furthermore, we retry the writing. 

I added some new tests regarding the schedule service. Are you interested in reviewing the changes?

File: engine/src/main/java/io/camunda/zeebe/streamprocessor/StreamProcessorContext.java
Comment (pihme) [inline engine/src/main/java/io/camunda/zeebe/streamprocessor/StreamProcessorContext.java:114]: :wrench: I think this could be package/default visibility; dito for setter

**Review by pihme — COMMENTED** [review]
Looks good to me

**Comment by pihme** [comment]
@Zelldon I cannot add myself as reviewer as I am the author. Nevertheless, I did the review and approve it.

**Comment by Zelldon** [comment]
:grin: thanks @pihme :bow:

**Review by ChrisKujawa — APPROVED** [review]
bors r+

**Pull request merged**
